tensorflow/c/kernels_experimental.cc

```diff
@@ -285,6 +288,97 @@ void TF_AssignUpdateVariable(TF_OpKernelContext* ctx, int input_index,
+std::string TemporaryVariableName(
+    const std::string& var_name,
+    const tensorflow::FrameAndIter& control_frame) {
+  if (control_frame.frame_id != tensorflow::kIllegalFrameId &&
+      control_frame.iter_id != tensorflow::kIllegalIterId) {
+    return tensorflow::strings::StrCat(var_name,
+                                       "/frame:", control_frame.frame_id,
+                                       "/iter:", control_frame.iter_id);
+  }
+  return var_name;
+}
+
+void TF_TemporaryVariable(TF_OpKernelContext* ctx, TF_DataType dtype,
+                          const int64_t* dims, int num_dims,
+                          TF_StringView* var_name,
+                          void (*allocFunc)(TF_OpKernelContext*, TF_Tensor*,
+                                            TF_DataType, const int64_t*, int,
+                                            TF_Status*),
+                          TF_Status* tf_status) {
+  auto* context = reinterpret_cast<::tensorflow::OpKernelContext*>(ctx);
+  tensorflow::ResourceMgr* rm = context->resource_manager();
+  OP_REQUIRES(context, rm,
+              tensorflow::errors::Internal("No per-step resource manager."));
+
+  std::string unique_name =
+      TemporaryVariableName(var_name->data, context->frame_iter());
+  auto* tmp_var = new TmpVar;
+  OP_REQUIRES(
+      context, tmp_var,
+      tensorflow::errors::ResourceExhausted("Could not allocate TmpVar."));
+  tmp_var->name = unique_name;
+
+  Status s;
+  std::unique_ptr<TF_Tensor, decltype(&TF_DeleteTensor)> tmp_var_tf(
+      tensorflow::TF_TensorFromTensor(tmp_var->val, &s), TF_DeleteTensor);
+  OP_REQUIRES_OK(context, s);
+  allocFunc(ctx, tmp_var_tf.get(), dtype, dims, num_dims, tf_status);
+  s = tensorflow::StatusFromTF_Status(tf_status);
+  if (!s.ok()) tmp_var->Unref();
+  OP_REQUIRES_OK(context, s);
+
+  OP_REQUIRES_OK(context, TF_TensorToTensor(tmp_var_tf.get(), &tmp_var->val));
+  OP_REQUIRES_OK(context,
+                 context->step_container()->Create(rm, unique_name, tmp_var));
+  context->set_output_ref(0, &tmp_var->mu, &tmp_var->val);
+
+  if (context->track_allocations()) {
+    context->record_persistent_memory_allocation(tmp_var->val.AllocatedBytes());
+  }
+
+  TF_SetStatus(tf_status, TF_OK, "");
+}
+
+void TF_DestroyTemporaryVariable(TF_OpKernelContext* ctx, const int index,
+                                 TF_StringView* var_name,
+                                 TF_Status* tf_status) {
+  auto* context = reinterpret_cast<::tensorflow::OpKernelContext*>(ctx);
+  if (!IsRefType(context->input_dtype(0))) {
+    tf_status->status =
+        InvalidArgument("TF_DestroyTemporaryVariable requires input is ref");
+    return;
+  }
+  Tensor tmpvar = context->mutable_input(0, false);
+  context->set_output(0, tmpvar);
+
+  tensorflow::ResourceMgr* rm = context->resource_manager();
+  OP_REQUIRES(context, rm,
+              tensorflow::errors::Internal("No per-step resource manager."));
```

Comment:
do you mind update to absl::Status? Thanks!

Reply:
done